File: 695. Max Area of Island/Python/Solution.py
class Solution:
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        steps = ((-1, 0), (1, 0), (0, -1), (0, 1))

        def dfs(grid, x, y, m, n):
            if x < 0 or x >= m or y < 0 or y >= n or grid[x][y] == 0:
                return 0
            count = 1
            grid[x][y] = 0
            for i, j in steps:
                count += dfs(grid, x + i, y + j, m, n)
            return count

        result = 0
        m = len(grid)
        n = len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    result = max(result, dfs(grid, i, j, m, n))
        return result

File: 695. Max Area of Island/C++/Solution.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>> &grid) {
        class Utils {
        public:
            int steps[4][2] = {{-1, 0},
                {1,  0},
                {0,  -1},
                {0,  1}
            };

            int dfs(vector<vector<int>> &grid, int x, int y, int m, int n) {
                if (x < 0 || x >= m || y < 0 || y >= n || grid[x][y] == 0) {
                    return 0;
                }
                int count = 1;
                grid[x][y] = 0;
                for (auto &step : steps) {
                    count += dfs(grid, x + step[0], y + step[1], m, n);
                }
                return count;
            }
        };
        int result = 0;
        int m = grid.size();
        int n = grid[0].size();
        Utils utils;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    result = max(result, utils.dfs(grid, i, j, m, n));
                }
            }
        }
        return result;
    }
};